Sources/Commands/Workspace.swift

```diff
@@ -400,6 +412,20 @@ public class Workspace {
         try saveState()
     }
 
+    /// Creates a clone at the provided version and pins the repository at that version.
+    ///
+    /// - Parameters:
+    ///   - repository: The repository to pin. It should not be already pinned.
+    ///   - at: The version to pin at.
+    ///   - reason: The optional reason for pinning.
+    /// - Throws: WorkspaceOperationError, PinOperationError
+    func pin(repository: RepositorySpecifier, at version: Version, reason: String? = nil) throws {
+        // Clone the version we want to pin.
+        _ = try clone(specifier: repository, version: version)
```

Comment:
We want to check that pin is valid to resolve, though, too right? So can't we let the full on machinery handle this?

The way I would imagine this working is a mechanism to try resolution + update checkouts with a provisional pin override -- if it works, then we commit that into the pins.